enhancements/network/persistent-ips-sdn-secondary-networks.md

```diff
@@ -807,3 +807,99 @@ allocations, it is probable the time it takes to provision a pod to grow.
 shortcomings - would be a major effort.
+
+### Separate IPAMLeases controller
+
+This alternative approach requires a separate controller to monitor both
+`PersistentIP` and `NetworkAttachmentDefinition` objects.
+
+KubeVirt would express its intent in having a `PersistentIP` for its VM
+interface (1 `PersistentIP` for each interface) by creating a CR; the new
+controller would learn about KubeVirt's desire, calculate an IP for it, and
+update the `PersistentIP` status with this information.
+
+KubeVirt VM templating engine would block the VM creating until it sees the
+persistent IP allocation available in the CR; once the IP is available in the
+CR, KubeVirt would be able to template the pod accordingly, requesting that IP
+in the interface's `NetworkSelectionElements`.
+
+This flow is portrayed in the following sequence diagram:
+
+```mermaid
+sequenceDiagram
+    actor user
+    user ->> kubevirt: Create(vm)
+    note over kubevirt, apiserver: leaseName = <vm>.<network>.<interface>
+    kubevirt->>apiserver: Create(ipamLease{name=leaseName})
+    apiserver->>ipamLeaseOperator: Reconcile(ipamLease)
+    Note right of ipamLeaseOperator: Status has no IPs
+    ipamLeaseOperator ->> ipamLeaseOperator: subnet = GetNAD(ipamLease).Subnet
+    ipamLeaseOperator ->> ipamLeaseOperator: ips = AllocateNextIPs(subnet)
+    ipamLeaseOperator ->> apiserver: UpdateStatus(ipamLease, ips)
+    loop until status has IPs
+        kubevirt->>apiserver: Get(ipamLease)
+    end
+    kubevirt ->> apiserver: Create(virtLauncher{ips: ipamLeases.IPs})
+```
+
+The flow for **stopping** the VM is quite simple - and similar to what we've
+seen so far; the user issues a VM stop command, which will trigger the launcher
+pod deletion. The persistent IP will **not** be deleted.
+
+```mermaid
+sequenceDiagram
+    actor user
+
+    user ->> kubevirt: Stop(vm)
+    kubevirt-->>user: Result
+    kubevirt->>apiserver: Delete(virtLauncher)
+```
+
+The flow for starting a stopped VM is again quite familiar; on a VM start event
+KubeVirt realizes there's already a persistent IP allocation for its interface,
+and will template the pod with it.
+
+```mermaid
+sequenceDiagram
+    actor user
+
+    user ->> kubevirt: Start(vm)
+    kubevirt->>apiserver: Get(ipamLease{name=vm.network.interface})
+    apiserver-->>kubevirt: ipamLease
+    kubevirt ->> apiserver: Create(virtLauncher{ips: ipamLeases.Status.IPs})
+```
+
+The flow for deleting a VM is similar to what we've seen so far; the Kubernetes
+garbage collector is the entity responsible for deleting the persistent IP
+allocation.
+
+```mermaid
+sequenceDiagram
+    actor user
+
+    user ->> apiserver: Delete(vm)
+    apiserver ->> gc: Delete(vm)
+    gc->>apiserver: Delete(virtLauncherPod)
+    apiserver-->>gc: Result
+    gc->>apiserver: Delete(ipamLease)
+    apiserver-->>gc: Result
+    apiserver->>ipamLeaseOperator: Reconcile(ipamLease)
+    ipamLeaseOperator ->> ipamLeaseOperator: DeallocateIPs(ipamLease.Status.IPs)
+```
+
```

Comment:
I am missing the sync when the IPAMLease operator get restarted or moved to another node or the like.

Reply:
Isn't that an implementation detail ? 

We care about the allocation flow, not exactly how will the IPs be reserved. I.e. I'm more concerned about explaining how will I block an IP from being allocated to other work-loads than explaining the inner working of the IP pool sync.

Reply:
Make sense, this enhancement is about impacts at current systems.

Reply:
closing / resolving